include/evmc/mocked_host.hpp

```diff
@@ -0,0 +1,320 @@
+            return 0;
+        return it->second.code.size();
+    }
+
+    /// Get the account's code hash (EVMC host method).
+    bytes32 get_code_hash(const address& addr) noexcept override
+    {
+        record_account_access(addr);
+        const auto it = accounts.find(addr);
+        if (it == accounts.end())
+            return {};
+        return it->second.codehash;
+    }
+
+    /// Copy the account's code to the given buffer (EVMC host method).
+    size_t copy_code(const address& addr,
+                     size_t code_offset,
+                     uint8_t* buffer_data,
+                     size_t buffer_size) noexcept override
+    {
+        record_account_access(addr);
+        const auto it = accounts.find(addr);
+        if (it == accounts.end())
+            return 0;
+
+        const auto& code = it->second.code;
+
+        if (code_offset >= code.size())
+            return 0;
+
+        const auto n = std::min(buffer_size, code.size() - code_offset);
+
+        if (n > 0)
+            std::copy_n(&code[code_offset], n, buffer_data);
+        return n;
+    }
+
+    /// Selfdestruct the account (EVMC host method).
+    void selfdestruct(const address& addr, const address& beneficiary) noexcept override
+    {
+        record_account_access(addr);
+        recorded_selfdestructs.push_back({addr, beneficiary});
+    }
+
+    /// Call/create other contract (EVMC host method).
+    result call(const evmc_message& msg) noexcept override
+    {
+        record_account_access(msg.destination);
+
+        if (recorded_calls.empty())
+        {
+            recorded_calls.reserve(max_recorded_calls);
+            m_recorded_calls_inputs.reserve(max_recorded_calls);  // Iterators will not invalidate.
+        }
+
+        if (recorded_calls.size() < max_recorded_calls)
+        {
+            recorded_calls.emplace_back(msg);
+            auto& call_msg = recorded_calls.back();
+            if (call_msg.input_size > 0)
+            {
+                m_recorded_calls_inputs.emplace_back(call_msg.input_data, call_msg.input_size);
+                const auto& input_copy = m_recorded_calls_inputs.back();
+                call_msg.input_data = input_copy.data();
+            }
+        }
+        return result{call_result};
+    }
+
+    /// Get transaction context (EVMC host method).
+    evmc_tx_context get_tx_context() noexcept override { return tx_context; }
+
+    /// Get the block header hash (EVMC host method).
+    bytes32 get_block_hash(int64_t block_number) noexcept override
+    {
+        recorded_blockhashes.emplace_back(block_number);
+        return block_hash;
+    }
+
+    /// Emit LOG (EVMC host method).
+    void emit_log(const address& addr,
+                  const uint8_t* data,
+                  size_t data_size,
+                  const bytes32 topics[],
+                  size_t topics_count) noexcept override
+    {
+        recorded_logs.push_back({addr, {data, data_size}, {}});
```

Comment:
`emplace_back`?

Reply:
Also probably you could just initialize topics vector inline with something like `std::vector<log_record>(&topics[0], topics + topics_count)`